### PR TITLE
Initial implementation of Grafana

### LINKED ISSUE #209: Grafana support
Grafana is one of the key components, together with ElasticSearch, that we (and many others) are using for monitoring and troubleshooting Kubernetes clusters. We should add support for Grafana to BKPR.

### PR DESCRIPTION
**Pull request by falfaro**
Initial implementation of Grafana in BKPR. It delegates authentication to oauth2_proxy and requires X-`Auth-Request-User` and `X-Auth-Request-Email` HTTP headers that identify the user.

Closes #209.

### CODE REVIEW

File: manifests/components/grafana.jsonnet [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  // OAuth2-related parameters
+  oauth2_scopes:: error "Missing OAuth2 scopes",
+  oauth2_auth_url:: error "Missing OAuth2 authentication URL",
+  oauth2_token_url:: error "Missing OAuth2 token URL",
+
+  // Google Bitnami oAuth secrets, sealed secrets under sre-kube-configs
+  google_oauth_secret: kube.Secret($.p + "grafana-google-oauth") + $.metadata {
+    data_+: {
+      "google-client-id": error "provided externally",
+      "google-client-secret": error "provided externally",
```
[the remark this answers is not in the document]
Comment (falfaro) [reply manifests/components/grafana.jsonnet]: I've placed all OAuth2-related parameters together under grafana.jsonnet at the beginning of the file.

File: docs/quickstart-gke.md [hunk 2]
```diff
@@ -51,6 +51,7 @@ In this section, you will deploy a Google Kubernetes Engine (GKE) cluster using
   8. Finally, add the following redirect URIs and hit __Create__ .
       + https://prometheus.{GCLOUD_DNS_ZONE}/oauth2/callback.
       + https://kibana.{GCLOUD_DNS_ZONE}/oauth2/callback.
+      + https://grafana.{GCLOUD_DNS_ZONE}/oauth2/callback.
```
[the remark this answers is not in the document]
Comment (falfaro) [reply docs/quickstart-gke.md]: No. GKE requires manual configuration for Oauth2. AKS however is set up automatically by BKPR. Look at pkg/aks/platform.go to see why :-)

File: manifests/components/grafana.jsonnet [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    },
+  },
+
+  grafana_admin_config: kube.Secret($.p + "grafana-admin-config") + $.metadata {
+    data_+: {
+      "grafana-admin-password": "admin",
```
[the remark this answers is not in the document]
Comment (falfaro) [reply manifests/components/grafana.jsonnet]: Yes, I could modify the part that generates the JSON configuration and generate a random secret and inject that into the manifest. What do you think?

File: manifests/platforms/gke.jsonnet [hunk 4]
```diff
@@ -42,6 +43,22 @@ local kibana = import "../components/kibana.jsonnet";
...
+      data_+: {
+        "google-client-id": $.config.oauthProxy.client_id,
+        "google-client-secret": $.config.oauthProxy.client_secret,
+      },
+    },
+    ingress+: {
+      host: "grafana." + $.external_dns_zone_name,
+    },
+  },
+
```
[the remark this answers is not in the document]
Comment (falfaro) [reply manifests/platforms/gke.jsonnet]: It's very short, and we are duplicating already some fragments like Prometheus ingress and Kibana ingress on both aks.jsonnet and gke.jsonnet.

File: manifests/components/grafana.jsonnet (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbianquetti-nami) [reply manifests/components/grafana.jsonnet]: 👍

File: manifests/platforms/gke.jsonnet (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbianquetti-nami) [reply manifests/platforms/gke.jsonnet]: ACK if you already are aware of this.

**Comment by dbarranco** [comment]
Before merging this PR I would add a ConfigMap with the configuration of the datasource of Prometheus using the internal Kubernetes cluster URL (prometheus.kubeprod.svc.cluster.local:port), so users will find pre-configured Grafana with that.
Hint: https://github.com/bitnami/sre-kube-configs/blob/master/grafana/provisioning/datasources/k.int/handcrafted/prometheus.yml

File: manifests/components/grafana.jsonnet [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+  // Default to "Editor".
+  // User can create/modify dashboards & alert rules, but cannot create/edit
+  // datasources nor invite new users
+  auto_role:: "Editor",
+
+  // Required to restrict login attempts to a particular domain
+  email_domain:: error "Missing e-mail domain",
+
+  // OAuth2-related parameters
```
[the remark this answers is not in the document]
Comment (falfaro) [reply manifests/components/grafana.jsonnet]: I've changed the PR to delegate authentication to oauth2_proxy.

**Comment by falfaro** [comment]
PTAL

**Comment by anguslees** [comment]
@falfaro: dbarranco is referring to the grafana "datasource" configuration.  It would be nice to ship it already configured to read from prometheus, since this is our use-case.

Eg see:
- https://github.com/coreos/prometheus-operator/blob/7f34279e7c69124a80248cc54cadef93fd1b2387/helm/grafana/templates/dashboards-configmap.yaml#L21
- http://docs.grafana.org/features/datasources/prometheus/

**Comment by dbarranco** [comment]
Hi, thanks for the explanation @anguslees.

Yeah, what would be nice is to have preconfigured Grafana with the internal URL of Prometheus (prometheus.kubeprod.svc.cluster.local:9090) so users will find Grafana automatically configured out of the box and can focus on creating dashboards.

**Comment by falfaro** [comment]
PTAL

File: manifests/components/grafana.jsonnet [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            paths: [
+              { path: "/", backend: $.svc.name_port },
+            ],
+          },
+        },
+      ],
+    },
+  },
+
+  // Generates YAML configuration file for under provisioning/datasources
```
Comment (jjo) [inline manifests/components/grafana.jsonnet]: Please fix the wording

File: manifests/components/grafana.jsonnet [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  p:: "",
+  metadata:: {
+    metadata+: {
+      namespace: "kubeprod",
+    },
+  },
+
+  prometheus:: error "No Prometheus service",
+
+  // Default to "Admin". See http://docs.grafana.org/permissions/overview/ for
```
Comment (jjo) [inline manifests/components/grafana.jsonnet]: I think we should better explain why are choosing `-#2` from below two reasonable choices:
-#1  Use `GF_SECURITY_ADMIN_USER` / `GF_SECURITY_ADMIN_PASSWORD`
   to explicitly set an `admin`,  then set `auto_role:: "Editor"`
-#2  All Org users are `Admin`

**Pull request merged**